### PR TITLE
feat: AM-7237 add traceability for organization license changes

### PR DESCRIPTION
**Pull request by wajb**
## :id: Reference related issue. 
[AM-7237](https://gravitee.atlassian.net/browse/AM-7237)

## :pencil2: A description of the changes proposed in the pull request
Trace organization license changes in management and gateway nodes with logging and audit events (managed cloud mode only).

## :memo: Test scenarios 
See jira.

## :computer: Add screenshots for UI
n/a

## :books: Any other comments that will help with documentation
n/a

## :man: :woman:@mentions of the person or team responsible for reviewing proposed changes
@gravitee-io/am

[AM-7237]: https://gravitee.atlassian.net/browse/AM-7237?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/LicenseServiceImpl.java [hunk 1]
```diff
@@ -65,40 +77,65 @@ public Maybe<License> findByReference(ReferenceType referenceType, String refere
...
+                .flatMap(existing -> {
+                    if (Objects.equals(existing.getLicense(), license)) {
+                        log.debug("License unchanged for referenceType={}, referenceId={}, skipping", referenceType, referenceId);
+                        return Maybe.just(existing);
+                    }
+                    final GraviteeLicense previous = describe(referenceType, referenceId, existing.getLicense());
+                    existing.setLicense(license);
+                    existing.setUpdatedAt(new Date());
+                    return licenseRepository.update(existing)
+                            .flatMap(updated -> emitEvent(referenceType, referenceId, Action.UPDATE).andThen(Single.just(updated)))
```
Comment (sclarkamlab) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/impl/LicenseServiceImpl.java:94]: The audit hooks sit downstream of `emitEvent`, so if the repo write succeeds but `eventService.create` fails, `reportChange` never runs and we record a FAILURE audit for a license that is already persisted. Same shape on the create and delete paths. Verified with a quick test, update succeeds then the event errors and the captured audit is `ORGANIZATION_LICENSE_UPDATED` / FAILURE. Could fix by auditing between the repository write and `emitEvent`, or by auditing the success and the notification failure separately.

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/LicenseServiceImpl.java [hunk 2]
```diff
@@ -65,40 +77,65 @@ public Maybe<License> findByReference(ReferenceType referenceType, String refere
...
-                        toCreate.setUpdatedAt(now);
-                        return licenseRepository.create(toCreate)
-                                .flatMap(created -> emitEvent(referenceType, referenceId, Action.CREATE).andThen(Single.just(created)));
-                    }));
-        });
+        return Maybe.<License>defer(() -> {
+                    validate(license);
+                    return licenseRepository.findById(referenceId, referenceType);
+                })
+                .doOnError(error -> reportFailure(EventType.ORGANIZATION_LICENSE_UPDATED, referenceType, referenceId, error))
```
Comment (sclarkamlab) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/impl/LicenseServiceImpl.java:84]: This fires before the create/update fork, so a failed validation or lookup on a first-time provisioning gets audited as `ORGANIZATION_LICENSE_UPDATED` for a license that never existed. The intent is unknowable up here, so maybe a neutral event type for pre-write failures rather than picking UPDATED.

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/LicenseServiceImpl.java [hunk 3]
```diff
@@ -117,4 +154,60 @@ private Completable emitEvent(ReferenceType referenceType, String referenceId, A
...
+                    .tier(license.getTier())
+                    .packs(license.getPacks())
+                    .features(license.getFeatures())
+                    .expiresAt(license.getExpirationDate())
+                    .isExpired(license.isExpired())
+                    .scope(license.getReferenceType())
+                    .build();
+        } catch (Exception e) {
+            log.warn("Cannot read license details for referenceType={}, referenceId={}", referenceType, referenceId, e);
+            return null;
```
Comment (sclarkamlab) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/impl/LicenseServiceImpl.java:177]: Returning null here means an undecodable license still audits as SUCCESS with no message at all, both sides of the diff are null so the builder skips the detail entirely (`LicenseServiceTest` pins this with `assertNull` on the message). The only trace of why is this warn log outside the trail. Worth putting a marker in the audit payload so an auditor can see the entitlements were unreadable rather than just missing.

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/LicenseServiceImpl.java (hunk 2 above)
Comment (wajb) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/LicenseServiceImpl.java:84]: Agreed it is unknowable. UPDATED was chosen as a reasonable default - with a comment in the test `createOrUpdateReportsFailureAuditWhenLookupFails`. I don't think we should add an in between type for the unlikely edge case. Wdyt?

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/LicenseServiceImpl.java (hunk 1 above)
Comment (wajb) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/LicenseServiceImpl.java:94]: This is fair, though it is the established pattern (see FactorServiceImpl, EmailTemplateServiceImpl, FormServiceImpl, ApplicationServiceImpl, AlertTriggerServiceImpl, etc.). This demonstrates failures during event emission is uncommon. Ok to follow convention here and follow up across all services if the need arises?

Comment (sclarkamlab) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/LicenseServiceImpl.java:94]: Yeah following the convention here is fine.

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/LicenseServiceImpl.java (hunk 2 above)
Comment (sclarkamlab) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/LicenseServiceImpl.java:84]: Yeah that is fair, UPDATED makes sense in that case then.

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/LicenseServiceImpl.java (hunk 3 above)
Comment (wajb) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/LicenseServiceImpl.java:177]: I created the sentinel license to use in this scenario but then determined it is unreachable in production. Invalid org licenses degrade to a static OSS license and the same `LicenseFactory` is used for parsing here. It would be a defensive fix but also dead code so I'm inclined to leave it out. Wdyt?

**Review by sclarkamlab — APPROVED** [review]
